third/nmh/man/mhmail.man
.\"
.\" %nmhwarning%
.\" $Id: mhmail.man,v 1.1.1.1 1999-02-07 18:14:21 danw Exp $
.\"
.\" include the -mh macro file
.so %etcdir%/tmac.h
.\"
.TH MHMAIL %manext1% MH.6.8 [%nmhversion%]
.SH NAME
mhmail \- send or read mail
.SH SYNOPSIS
.in +.5i
.ti -.5i
mhmail
\%[
addrs\ ... 
\%[\-body\ text]
\%[\-cc\ addrs\ ...]
.br
\%[\-from\ addr]
\%[\-subject subject]]
.br
\%[\-version]
\%[\-help]
.in -.5i
.SH DESCRIPTION
\fImhmail\fR is intended as a replacement for the standard Berkeley
mail program (\fImail\fR(1) or \fImailx\fR(1)), which is compatible
with \fInmh\fR.  This program is intended for the use of programs such
as \fIcron\fR(1), which expect to send mail automatically to various
users.  It is also used by various \fInmh\fR commands to mail various
error notifications.  Although \fImhmail\fR can be used interactively,
it is recommended that \fIcomp\fR(1) and \fIsend\fR(1) be used instead
to send messages.

When invoked without arguments, it simply invokes \fIinc\fR(1) to
incorporate new messages from the user's maildrop.  When one or more users
is specified, a message is read from the standard input and spooled to
a temporary file.  \fImhmail\fR then invokes \fIpost\fR(8) with the
name of the temporary file as its argument to deliver the message to
the specified user.

The `\-subject\ subject' switch can be used to specify the
\*(lqSubject:\*(rq field of the message.

By default, \fImhmail\fR will read the message to be sent from the
standard input.  You can specify the text of the message at the command
line with the `\-body\ text' switch.  If the standard input has zero
length, \fImhmail\fR will not send the message.  You can use the switch
`\-body\ ""' to force an empty message.

Normally, addresses appearing as arguments are put in the \*(lqTo:\*(rq
field.  If the `\-cc' switch is used, all addresses following it are
placed in the \*(lqcc:\*(rq field.

By using `\-from\ addr', you can specify the \*(lqFrom:\*(rq header of
the draft.  Naturally, \fIpost\fR will fill\-in the \*(lqSender:\*(rq
header correctly.
.Fi
^%bindir%/inc~^Program to incorporate maildrop into folder
^%libdir%/post~^Program to deliver a message
^/tmp/mhmail*~^Temporary copy of message
.Pr
None
.Sa
inc(1), post(8)
.De
None
.Co
If \fIinc\fR is invoked, then \fIinc\fR's context changes occur.
.En


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/nmh/uip/mhmail.c

/*
 * mhmail.c -- simple mail program
 *
 * $Id: mhmail.c,v 1.1.1.1 1999-02-07 18:14:14 danw Exp $
 */

#include <h/mh.h>
#include <h/signals.h>
#include <signal.h>

static struct swit switches[] = {
#define	BODYSW             0
    { "body text", 0 },
#define	CCSW               1
    { "cc addrs ...", 0 },
#define	FROMSW             2
    { "from addr", 0 },
#define	SUBJSW             3
    { "subject text", 0 },
#define VERSIONSW          4
    { "version", 0 },
#define	HELPSW             5
    { "help", 4 },
#define	RESNDSW            6
    { "resent", -6 },
#define	QUEUESW            7
    { "queued", -6 },
    { NULL, 0 }
};

static char tmpfil[BUFSIZ];

/*
 * static prototypes
 */
static RETSIGTYPE intrser (int);


int
main (int argc, char **argv)
{
    pid_t child_id;
    int status, i, iscc = 0, nvec;
    int queued = 0, resent = 0, somebody;
    char *cp, *tolist = NULL, *cclist = NULL, *subject = NULL;
    char *from = NULL, *body = NULL, **argp, **arguments;
    char *vec[5], buf[BUFSIZ];
    FILE *out;

#ifdef LOCALE
    setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');

    /* foil search of user profile/context */
    if (context_foil (NULL) == -1)
	done (1);

    /* If no arguments, just incorporate new mail */
    if (argc == 1) {
	execlp (incproc, r1bindex (incproc, '/'), NULL);
	adios (incproc, "unable to exec");
    }

    arguments = getarguments (invo_name, argc, argv, 0);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    adios (NULL, "-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [addrs ... [switches]]",
			invo_name);
		    print_help (buf, switches, 0);
		    done (1);
		case VERSIONSW:
		    print_version(invo_name);
		    done (1);

		case FROMSW: 
		    if (!(from = *argp++) || *from == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;

		case BODYSW: 
		    if (!(body = *argp++) || *body == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;

		case CCSW: 
		    iscc++;
		    continue;

		case SUBJSW: 
		    if (!(subject = *argp++) || *subject == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;

		case RESNDSW: 
		    resent++;
		    continue;

		case QUEUESW: 
		    queued++;
		    continue;
	    }
	}
	if (iscc)
	    cclist = cclist ? add (cp, add (", ", cclist)) : getcpy (cp);
	else
	    tolist = tolist ? add (cp, add (", ", tolist)) : getcpy (cp);
    }

    if (tolist == NULL)
	adios (NULL, "usage: %s addrs ... [switches]", invo_name);
    strncpy (tmpfil, m_tmpfil (invo_name), sizeof(tmpfil));
    if ((out = fopen (tmpfil, "w")) == NULL)
	adios (tmpfil, "unable to write");
    chmod (tmpfil, 0600);

    SIGNAL2 (SIGINT, intrser);

    fprintf (out, "%sTo: %s\n", resent ? "Resent-" : "", tolist);
    if (cclist)
	fprintf (out, "%scc: %s\n", resent ? "Resent-" : "", cclist);
    if (subject)
	fprintf (out, "%sSubject: %s\n", resent ? "Resent-" : "", subject);
    if (from)
	fprintf (out, "%sFrom: %s\n", resent ? "Resent-" : "", from);
    if (!resent)
	fputs ("\n", out);

    if (body) {
	fprintf (out, "%s", body);
	if (*body && *(body + strlen (body) - 1) != '\n')
	    fputs ("\n", out);
    } else {
	for (somebody = 0;
		(i = fread (buf, sizeof(*buf), sizeof(buf), stdin)) > 0;
		somebody++)
	    if (fwrite (buf, sizeof(*buf), i, out) != i)
		adios (tmpfil, "error writing");
	if (!somebody) {
	    unlink (tmpfil);
	    done (1);
	}
    }
    fclose (out);

    nvec = 0;
    vec[nvec++] = r1bindex (postproc, '/');
    vec[nvec++] = tmpfil;
    if (resent)
	vec[nvec++] = "-dist";
    if (queued)
	vec[nvec++] = "-queued";
    vec[nvec] = NULL;

    for (i = 0; (child_id = fork()) == NOTOK && i < 5; i++)
	sleep (5);

    if (child_id == NOTOK) {
	/* report failure and then send it */
	admonish (NULL, "unable to fork");
    } else if (child_id) {
	/* parent process */
	if ((status = pidXwait(child_id, postproc))) {
	    fprintf (stderr, "Letter saved in dead.letter\n");
	    execl ("/bin/mv", "mv", tmpfil, "dead.letter", NULL);
	    execl ("/usr/bin/mv", "mv", tmpfil, "dead.letter", NULL);
	    perror ("mv");
	    _exit (-1);
	}
	unlink (tmpfil);
	done (status ? 1 : 0);
    } else {
	/* child process */
	execvp (postproc, vec);
	fprintf (stderr, "unable to exec ");
	perror (postproc);
	_exit (-1);
    }
}


static RETSIGTYPE
intrser (int i)
{
#ifndef RELIABLE_SIGNALS
    if (i)
	SIGNAL (i, SIG_IGN);
#endif

    unlink (tmpfil);
    done (i != 0 ? 1 : 0);
}

